AMD: libmetal demo - add baremetal support

bentheredonethat opened the pull request:
Hi @arnopo  this is based off the approved PR https://github.com/OpenAMP/openamp-system-reference/pull/82 

changes here are:

1. refactor to make addition of new demos more scalable
3. add baremetal build support

edmooring reviewed (approved):
I don't see any issues.

bentheredonethat commented:
please hold for now - i need to do additional testing on this with latest upstream

bentheredonethat committed:
examples: libmetal: amd_rpu: Add Baremetal Port

Add port that implements, main routine, suspend/resume
and wiring in interrupts.

Signed-off-by: Ben Levinsky <ben.levinsky@amd.com>

bentheredonethat commented:
updated and tested now @arnopo

bentheredonethat committed:
examples: libmetal: amd_rpu: fix linker script

Ensure that RPU Firmware will boot using FreeRTOS or Baremetal
system by making sure linker script is proper.

Signed-off-by: Ben Levinsky <ben.levinsky@amd.com>

bentheredonethat committed:
examples: libmetal: amd_rpu: Set size for shared memory device based on macro

Before the size was hard-coded. Now let macro definition SHM_SIZE
set shared memory region size

Signed-off-by: Ben Levinsky <ben.levinsky@amd.com>

bentheredonethat committed:
examples: libmetal: amd_rpu: IRQ Handler now to handle based on registed IRQ

As the IPI_IRQ_VECT_ID system interrupt is what is used to map
the internal IRQ handler, pass this in to metal_irq_handle()
call.

Signed-off-by: Ben Levinsky <ben.levinsky@amd.com>

bentheredonethat committed:
examples: libmetal: amd_rpu: Ensure IRQ setup is clean

Wipe pre-existing IRQ kicks before register and enable
of handler for IRQ

Signed-off-by: Ben Levinsky <ben.levinsky@amd.com>

bentheredonethat committed:
examples: libmetal: amd_rpu: Update log for role

As the role for this C file is remote, ensure
metal_err logging reflects this.

Signed-off-by: Ben Levinsky <ben.levinsky@amd.com>

tnmysh commented on examples/libmetal/machine/remote/amd_rpu/system/baremetal/CMakeLists.txt:
Convention for "baremetal" in this project everywhere is "generic". We have designed toolchains accordingly, so just for the sake of the convention, can we rename "baremetal" -> "generic" ?

bentheredonethat replied:
i have done this in latest branch

bentheredonethat committed:
examples: libmetal: amd_rpu: Add build dependencies

This was skipped because of .gitignore - add this for compilation

Signed-off-by: Ben Levinsky <ben.levinsky@amd.com>

bentheredonethat commented on examples/libmetal/machine/remote/amd_rpu/system/baremetal/gic_init.c:
```diff
@@ -0,0 +1,62 @@
+
+static XIpiPsu IpiInst;
+
+void xlnx_irq_isr(void *arg)
```
i have done this in latest branch

bentheredonethat commented on examples/libmetal/machine/remote/amd_rpu/lscript_sdt.ld:
```diff
@@ -22,12 +22,12 @@ MEMORY
    psu_r5_atcm_MEM_0 : ORIGIN = 0x0, LENGTH = 0x10000
    psu_r5_btcm_MEM_0 : ORIGIN = 0x20000, LENGTH = 0x10000
-   psu_r5_ddr_0_MEM_0 : ORIGIN = 0x3ed00000, LENGTH = 0x80000
+   psu_r5_ddr_0_MEM_0 : ORIGIN = 0x9800100, LENGTH = 0x5ff00
```
@tnmysh  this will eventually become data driven - we have linker .in files that, when configure will ensure the data fits. in the future this problem will be moot.

bentheredonethat committed:
libmetal: amd_rpu: use libmetal logging framework

Same as legacy_apps, define and use platform specific logging handler.

Signed-off-by: Ben Levinsky <ben.levinsky@amd.com>

tnmysh commented on examples/libmetal/machine/remote/amd_rpu/platform_init.c:
```diff
@@ -214,7 +214,7 @@ int platform_register_metal_device(void)
 	     i++) {
 		dev = &metal_dev_table[i];
-		metal_info("HOST: registering: %d, name=%s\n", i, dev->name);
+		metal_info("REMOTE: registering: %d, name=%s\n", i, dev->name);
```
struct metal_init_params metal_param = XLNX_PLATFORM_METAL_INIT_PARAMS; 

Please add above where metal_param is set in platform_init.

bentheredonethat replied:
@tnmysh  addressed this one

tnmysh commented on examples/libmetal/machine/remote/amd_rpu/system/generic/gic_init.c:
```diff
@@ -0,0 +1,62 @@
+extern void xlnx_irq_isr(void *arg);
+
+/* Interrupt Controller setup */
+int init_irq(void)
```
Right now this init_irq interface does not accept any arguments, but would like to change it in a way where it accepts interrupt number and interrupt handler as an argument.

https://github.com/OpenAMP/openamp-system-reference/blob/b73f5d7ca3e67b3cde6e8f1283a62118f83a1c70/examples/legacy_apps/machine/zynqmp_r5/generic/gic_init.c#L29

I had done this here. Again, not critical if demo is working. We can change it later.

tnmysh reviewed (approved):
LGTM.

arnopo reviewed (commented):
One typo to fix then ready to be merged

bentheredonethat commented on examples/libmetal/machine/remote/amd_rpu/platform_init.c:
```diff
@@ -276,29 +292,29 @@ void close_metal_devices(void)
-	if (!IpiCfgPtr) {
+	ipi_cfg_ptr = XIpiPsu_LookupConfig(XPAR_XIPIPSU_0_BASEADDR);
+	if (!ipi_cfg_ptr) {
 		metal_err("HOST: %s ERROR in getting CfgPtr\n", __func__);
```
apologies I think that was lost on old branch @arnopo
